### PR TITLE
feat: add invocation state; apply to all events

### PR DESCRIPTION
**Pull request by notowen333**
## Description

Adds `invocationState`: a per-invocation, mutable `Record<string, unknown>` that callers pass into `agent.invoke()` / `agent.stream()`, threaded through every hookable event and every tool, and returned on `AgentResult.invocationState`. Ports Python's `request_state` concept to TS, letting callers thread request-scoped context (`userId`, `requestId`, `traceId`, custom counters) through a single run without polluting model context or leaking into durable agent state.

Part of this PR adds invocationState to all of the events in the SDK. We are aligned that we own these events and prefer to make breaking changes to custom plugins by adding events instead of introducing a sea of ?

```typescript
// Pass state in, read it back on the result
const result = await agent.invoke('Hi', {
  invocationState: { userId: 'u-1', requestId: 'r-42' },
})
result.invocationState.userId // 'u-1' — same object round-trips

// Hook callbacks can read and mutate
agent.addHook(BeforeModelCallEvent, (event) => {
  event.invocationState.modelCalls = (event.invocationState.modelCalls ?? 0) + 1
})

// Tools see it via ToolContext
tool({
  name: 'getUser',
  callback: (_input, ctx) => `User: ${ctx.invocationState.userId}`,
})

// Forwarded through multi-agent orchestration and agent-as-tool composition
await graph.invoke('hello', { invocationState: { traceId: 't-1' } })
const writer = new Agent({ tools: [researcher.asTool()] })
await writer.invoke('...', { invocationState }) // flows into researcher
```

**Shape**: `type InvocationState = Record<string, unknown>`. One object per invocation, shared by reference across every hook event, every tool, and every recursive loop cycle. Mutations are visible to subsequent hooks and tools. Defaults to `{}` when omitted. Excluded from `toJSON()` because values are caller-owned and may not be serializable. Distinct from `agent.appState` (durable, JSON-validated, deep-copied, session-persistent).

**Key design decisions diverging from Python:**

1. **Flattened Python's two-layer shape into one.** Python exposes `invocation_state` (outer) containing `request_state` (inner, returned on the result). That split exists because Python also writes SDK-internal keys into `invocation_state` — `"agent"`, `"event_loop_cycle_id"`, cycle traces, spans, etc. — so `request_state` got carved out as "this is yours, not the SDK's". TS has typed homes for all of those (`ToolContext.agent`, `Tracer`, `Meter`, typed hook event fields), so the pollution problem doesn't exist and the carveout isn't needed. One flat `invocationState`, zero SDK-written keys, the entire key space belongs to the caller.

2. **`invocationState` on every event that fires during an invocation, not just lifecycle ones.** The first draft only put it on lifecycle and state-change events (Before/After Invocation/Model/Tools/ToolCall, MessageAdded), mirroring Python's hook event surface. Python doesn't expose data/streaming events as hook events at all — callers can't subscribe to per-chunk model streaming in Python. TS has a richer hookable surface, and the callbacks that most need request-identity context are exactly the streaming ones: token-fanning to per-user UI panels, per-chunk distributed-trace spans, per-content-block audit logging. Omitting `invocationState` from those events would push callers into closure workarounds for the feature's main use case. Final rule: every hookable event that fires during an invocation carries `invocationState`. Only `InitializedEvent` and `MultiAgentInitializedEvent` don't — they fire at construction, before any invocation exists.

3. **No `stopEventLoop` control flag.** Python reserves `request_state["stop_event_loop"] = True` as an in-loop stop signal (experimental, bidi-only). TS already has `agent.cancel()` / `InvokeOptions.cancelSignal` — structured, propagates into in-flight model streams, produces `stopReason: 'cancelled'`. Adding a magic dict key would give two ways to stop with the lossier one polluting `invocationState`. If TS later needs a graceful "stop after this cycle" signal, it gets its own typed mechanism.

4. **`Record<string, unknown>`, not generic `Agent<TState>`.** Matches Python's untyped-bag contract. A generic `TState` would virus through `Agent`, `AgentResult`, `InvokeOptions`, every hook event, `ToolContext`, `Node`, `MultiAgent` — large surface change for a feature most callers won't parameterize. Adding a generic default later is non-breaking.

## Related Issues

#772 

## Documentation PR

will follow up 

## Type of Change

Breaking change

## Testing

How have you tested the change?

- [x] I ran `npm run check`

Added dedicated tests for: round-trip, reference identity across all hook events, hook mutation visibility, multi-cycle persistence through tool-use loops, retry-path reference preservation (`AfterModelCallEvent.retry` and `AfterToolCallEvent.retry`), isolation from `appState`, cross-invocation isolation, tool access via `ToolContext`, agent-as-tool forwarding, Graph and Swarm multi-node forwarding, and an identity sweep across every orchestrator-and-node-level event in a Graph run.

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

## Breaking Changes

Constructors for hook events and `AgentResult` now require `invocationState` in their data object. External code constructing these directly (custom plugins, test fixtures) will fail to compile until updated. Normal usage — invoking agents, registering hooks, writing tools — is unaffected.

### Migration

```typescript
// Before
new BeforeInvocationEvent({ agent })
new AgentResult({ stopReason, lastMessage, metrics })

// After
new BeforeInvocationEvent({ agent, invocationState })
new AgentResult({ stopReason, lastMessage, metrics, invocationState })
```

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: strands-ts/src/a2a/a2a-agent.ts [hunk 1]
```diff
@@ -111,12 +111,16 @@ export class A2AAgent implements InvokableAgent {
...
+   *   provided, it is returned on the resulting `AgentResult`. The remote
+   *   agent runs in another process and cannot read or mutate it. Other
+   *   fields on `options` are ignored.
    * @returns Async generator that yields AgentStreamEvent objects and returns AgentResult
    */
-  async *stream(args: InvokeArgs, _options?: InvokeOptions): AsyncGenerator<A2AStreamEvent, AgentResult, undefined> {
+  async *stream(args: InvokeArgs, options?: InvokeOptions): AsyncGenerator<A2AStreamEvent, AgentResult, undefined> {
     const client = await this._getClient()
     const text = this._extractTextFromArgs(args)
+    const invocationState = options?.invocationState ?? {}
```
Comment (mkmeral) [inline strands-ts/src/a2a/a2a-agent.ts:123]: in python a2a agent, we also passing metadata to a2a from invocation state. we should have the same,allows for sending headers, etc.

**Comment by agent-of-mkmeral** [comment]
**TL;DR**: FAIL — 2 issues found. Adversarial tested `invocationState` threading across all 38 changed files with 32 targeted tests. toJSON exclusion is solid; two data-flow gaps confirmed.

<details>
<summary>Adversarial Testing Report</summary>

**Scope**: `invocationState` propagation through agent loop, A2A executor, multiagent nodes (Graph/Swarm), agent-as-tool, all hook events, and serialization boundaries.

**Baseline**: 2037/2041 existing tests pass (4 pre-existing telemetry config failures, unrelated to this PR).

**Adversarial tests written**: 32
**Adversarial tests passing**: 32
**Findings**: 2

| # | Category | Severity | Description |
|---|----------|----------|-------------|
| 1 | Contract Violation | High | `A2AExecutor` does not pass `invocationState` to the agent — parity gap with Python |
| 2 | Unhandled Edge Case | Medium | Double-resolution of `invocationState` in `AgentNode.handle()` / `MultiAgentNode.handle()` creates divergent references when called directly |

<details>
<summary>Finding 1 — Contract Violation (High): A2AExecutor does not forward invocationState to agent</summary>

**Location**: `src/a2a/executor.ts:83`

**Reproduction**:
```typescript
// executor.ts line 83:
const stream = this._agent.stream(contentBlocks)
// ^ No InvokeOptions passed — invocationState is undefined
```

Adversarial test proves it:
```typescript
// After executor.execute(context, eventBus):
const callArgs = (spy.agent.stream as any).mock.calls[0]
const passedOptions = callArgs[1] as InvokeOptions | undefined
expect(passedOptions?.invocationState).toBeUndefined() // ✅ PASSES — confirms the bug
```

**Observed behavior**: `A2AExecutor.execute()` calls `this._agent.stream(contentBlocks)` with no second argument. The agent's internal `_stream()` method then defaults `invocationState` to `{}`. Hooks and tools running inside the A2A-served agent receive an empty `invocationState` with no way to access the A2A `RequestContext` (taskId, contextId, user message metadata).

**Expected behavior** (Python parity):
```python
# Python src/strands/multiagent/a2a/executor.py:133-136
invocation_state: dict[str, Any] = {"a2a_request_context": context}
async for event in self.agent.stream_async(content_blocks, invocation_state=invocation_state)
```

**Impact**: Any hook or tool running inside an A2A-served agent that needs to correlate with the A2A request (e.g., logging taskId, per-request metrics, audit trails) has no access to the request context. This is the primary use case the issue author asked about.

**Fix**:
```typescript
// In executor.ts:
const stream = this._agent.stream(contentBlocks, {
  invocationState: { a2aRequestContext: context },
})
```

**Test file**: `src/a2a/__tests__/test_adversarial_executor_invocation_state.test.ts` (2 tests)

</details>

<details>
<summary>Finding 2 — Unhandled Edge Case (Medium): Double-resolution creates divergent references in node handle()</summary>

**Location**: `src/multiagent/nodes.ts:191` (AgentNode) and `src/multiagent/nodes.ts:278` (MultiAgentNode)

**Reproduction**:
```typescript
// Node.stream() resolves invocationState at line 89:
const invocationState: InvocationState = options?.invocationState ?? {}
// Then passes it to handle() via resolvedOptions

// AgentNode.handle() resolves AGAIN at line 191:
const invocationState: InvocationState = options?.invocationState ?? {}
// When called directly (not via stream()), this creates a DIFFERENT {} object
```

Adversarial test proves divergence:
```typescript
const seenStates: InvocationState[] = []
// Call handle() directly twice without invocationState
const gen1 = node.handle('test', state, undefined)
// ... consume gen1 ...
const gen2 = node.handle('test', state, undefined)
// ... consume gen2 ...

// Each call created its OWN separate {} object:
expect(seenStates[0]).not.toBe(seenStates[1]) // ✅ PASSES — different references
```

**Observed behavior**: When `handle()` is called directly (which is a documented valid call path per the comment "makes handle() safe to call directly"), it creates its own `{}` for invocationState. If `stream()` calls `handle()`, `stream()`'s resolved state is passed through correctly. But the dual resolution means:
1. Direct callers of `handle()` get isolated (non-shared) state objects
2. If a custom node subclass calls `handle()` directly without forwarding its own `invocationState`, mutations in one path aren't visible in the other

**Expected behavior**: Single resolution point. Either:
- Remove the `?? {}` from `handle()` and document that callers must provide `invocationState` (TypeScript's type system can enforce this)
- Or keep handle()'s resolution but document that direct callers get an independent state object

**Impact**: Medium. In practice, `handle()` is called via `stream()` which passes the resolved state through. The issue only manifests if someone calls `handle()` directly without providing `invocationState`. But the pattern is confusing and creates a footgun for custom node implementations.

**Test file**: `src/multiagent/__tests__/test_adversarial_nodes_double_resolution.test.ts` (3 tests)

</details>

<details>
<summary>What Survived (30 tests passed)</summary>

**toJSON exclusion verified exhaustively (27 tests)**:
- All 15 agent hook events (`BeforeInvocationEvent`, `AfterInvocationEvent`, `BeforeModelCallEvent`, `AfterModelCallEvent`, `BeforeToolCallEvent`, `AfterToolCallEvent`, `BeforeToolsEvent`, `AfterToolsEvent`, `MessageAddedEvent`, `ModelStreamUpdateEvent`, `ContentBlockEvent`, `ModelMessageEvent`, `ToolResultEvent`, `ToolStreamUpdateEvent`, `AgentResultEvent`) correctly exclude `invocationState` from `toJSON()`
- All 9 multiagent events correctly exclude `invocationState` from `toJSON()`
- `AgentResult.toJSON()` correctly excludes `invocationState`
- Secret data in `invocationState` (API keys, passwords) does NOT leak through `JSON.stringify()`
- `MultiAgentInitializedEvent` correctly has no `invocationState` (fires before any invocation)

**Reference identity (3 tests)**:
- Through the `stream()` → `handle()` path, `NodeResultEvent` and inner agent events share the same `invocationState` reference ✅
- Agent-as-tool forwarding works correctly (covered by existing PR tests)
- Graph/Swarm threading works correctly (covered by existing PR tests)

</details>

</details>

---
🤖 *AI agent response. [Strands Agents](https://github.com/strands-agents). Feedback welcome!*

**Review by mkmeral — COMMENTED** [review]
also check out the comment from my agent

File: strands-ts/src/types/agent.ts [hunk 2]
```diff
@@ -35,6 +35,26 @@ import { AgentMetrics } from '../telemetry/meter.js'
...
+ *
+ * Distinct from {@link LocalAgent.appState}: `appState` is durable across
+ * invocations, JSON-serializable, and deep-copied. `invocationState` is
+ * ephemeral and accepts arbitrary values.
+ *
+ * Excluded from `toJSON()` on {@link AgentResult} and all hook events because
+ * values may not be serializable; callers produce a serialized form explicitly
+ * if needed.
+ */
+export type InvocationState = Record<string, unknown>
```
[the remark this answers is not in the document]
Comment (notowen333) [reply strands-ts/src/types/agent.ts:58]: Applied the `needs-api-review` label on the PR.

File: strands-ts/src/multiagent/nodes.ts [hunk 3]
```diff
@@ -166,12 +184,19 @@ export class AgentNode extends Node {
     state: MultiAgentState,
     options?: NodeInputOptions
   ): AsyncGenerator<MultiAgentStreamEvent, NodeResultUpdate, undefined> {
+    // Resolve invocationState once so every NodeStreamUpdateEvent and the
+    // inner agent's invocation share one reference. Node.stream normally
+    // passes this through already, but resolving here makes handle() safe to
+    // call directly.
+    const invocationState: InvocationState = options?.invocationState ?? {}
```
[the remark this answers is not in the document]
Comment (notowen333) [reply strands-ts/src/multiagent/nodes.ts:189]: Kept the `?? {}` fallback and tightened the comment to defend the pattern. `handle()` is public API — direct callers legitimately get per-call state. `Node.stream()` still threads one resolved reference through `resolvedOptions`, so the on-the-stream path is identity-preserving. The "divergent references" scenario requires mixing direct `handle()` calls with `stream()` in the same logical run, which is incoherent usage.

File: strands-ts/src/multiagent/nodes.ts [hunk 4]
```diff
@@ -238,15 +264,20 @@ export class MultiAgentNode extends Node {
...
   async *handle(
     input: MultiAgentInput,
     state: MultiAgentState,
-    _options?: NodeInputOptions
+    options?: NodeInputOptions
   ): AsyncGenerator<MultiAgentStreamEvent, NodeResultUpdate, undefined> {
-    const gen = this._orchestrator.stream(input)
+    // Resolve once so every wrapped NodeStreamUpdateEvent and the nested
+    // orchestrator's invocation share the same reference.
+    const invocationState: InvocationState = options?.invocationState ?? {}
```
[the remark this answers is not in the document]
Comment (notowen333) [reply strands-ts/src/multiagent/nodes.ts:276]: Same resolution as the `AgentNode.handle()` thread — kept the fallback and tightened the comment. `Node.stream()` preserves reference identity via `resolvedOptions`; direct callers of `handle()` get per-call state, which is semantically correct.

File: strands-ts/src/agent/__tests__/agent-as-tool.invocation-state.test.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    const outerModel = new MockMessageModel()
+      .addTurn([{ type: 'toolUseBlock', name: 'inner', toolUseId: 'tu-1', input: { input: 'hi' } }])
+      .addTurn({ type: 'textBlock', text: 'outer-done' })
+    const outer = new Agent({ model: outerModel, tools: [inner.asTool()] })
+
+    const result = await outer.invoke('run inner', { invocationState: { userId: 'u-1' } })
+
+    expect(innerSawState?.userId).toBe('u-1')
+    expect(result.invocationState.innerTouched).toBe(true)
```
[the remark this answers is not in the document]
Comment (notowen333) [reply strands-ts/src/agent/__tests__/agent-as-tool.invocation-state.test.ts]: Switched to full-object `toEqual`:
```ts
expect(innerSawState).toEqual({ userId: 'u-1', innerTouched: true })
expect(result.invocationState).toEqual({ userId: 'u-1', innerTouched: true })
```

File: strands-ts/src/agent/__tests__/agent.invocation-state.test.ts [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+      }
+
+      const model = new MockMessageModel()
+        .addTurn([{ type: 'toolUseBlock', name: 'writer', toolUseId: 'tu-1', input: {} }])
+        .addTurn({ type: 'textBlock', text: 'Done' })
+      const agent = new Agent({ model, tools: [tool] })
+
+      const result = await agent.invoke('Run writer', { invocationState: { userId: 'u-42' } })
+
+      expect(result.invocationState.callCount).toBe(1)
```
[the remark this answers is not in the document]
Comment (notowen333) [reply strands-ts/src/agent/__tests__/agent.invocation-state.test.ts]: Swept the file for per-field assertions and converted them to full-object `toEqual`. The tool-access test now uses:
```ts
expect(result.invocationState).toEqual({ userId: 'u-42', callCount: 1, lastToolSeenUserId: 'u-42' })
```
Reference-identity `.toBe` assertions were kept where they test identity rather than content.

File: strands-ts/src/a2a/a2a-agent.ts (hunk 1 above)
Comment (notowen333) [reply strands-ts/src/a2a/a2a-agent.ts:123]: Done — `A2AExecutor.execute()` now forwards the `RequestContext` under the reserved key `a2aRequestContext` (camelCase for TS convention):

```ts
const stream = this._agent.stream(contentBlocks, {
  invocationState: { a2aRequestContext: context },
})
```

Added a dedicated test verifying the key lands on the child agent's `invocationState`, updated `A2AExecutor` class JSDoc with an "Invocation state" section documenting the reserved key, and updated the `InvocationState` type doc to note that transport bridges may populate reserved keys.

File: strands-ts/src/agent/agent.ts [hunk 7]
```diff
@@ -604,22 +607,27 @@ export class Agent implements LocalAgent, InvokableAgent {
...
+    // Resolve per-invocation state once. The same object is threaded through
+    // every lifecycle hook event, every tool context, and is surfaced on the
+    // AgentResult. Mutations by hooks/tools are visible across all recursive
+    // agent loop cycles within this invocation.
+    const invocationState: InvocationState = options?.invocationState ?? {}
+
+    const beforeInvocationEvent = new BeforeInvocationEvent({ agent: this, invocationState })
     yield beforeInvocationEvent
 
     if (beforeInvocationEvent.cancel) {
```
[the remark this answers is not in the document]
Comment (notowen333) [reply strands-ts/src/agent/agent.ts:644]: The cancel field on before invocation event is from rebasing

**Comment by agent-of-mkmeral** [comment]
**TL;DR**: PASS — 0 issues found. The changes since our last adversarial test (Apr 24) correctly address both prior findings. 16 new adversarial tests all pass.

<details>
<summary>Adversarial Testing Report (re-test after fixes)</summary>

**Scope**: Re-validated the `fix: address PR comments` commit (`d2d2f94`) which addressed our two prior findings + reviewer feedback.

**Baseline**: 2141/2145 existing tests pass (4 pre-existing telemetry config failures, unrelated to this PR).

**Adversarial tests written**: 16
**Adversarial tests passing**: 16
**Findings**: 0

### Prior Findings — Verified Fixed

| # | Original Finding | Status | Verification |
|---|-----------------|--------|-------------|
| 1 | A2AExecutor doesn't pass invocationState | ✅ Fixed | `executor.ts:94` now passes `{ invocationState: { a2aRequestContext: context } }`. Dedicated test added. JSDoc documents the reserved key. |
| 2 | Double-resolution in node handle() | ✅ Documented | Comments updated to clarify `handle()` is public API — direct callers get per-call state. Design decision is intentional and well-explained. |

### What the Fix Commit Changed
- **A2AExecutor**: Forwards `RequestContext` under `a2aRequestContext` key (Python parity ✅)
- **Nodes.ts**: Comments clarified for the double-resolution pattern
- **Tests**: Per-field assertions → full-object `toEqual` (reviewer feedback)
- **Fixture**: `expectAgentResult` now validates `invocationState`
- **Type doc**: `InvocationState` JSDoc updated to mention transport bridge reserved keys

### Adversarial Tests Written (16 total, all pass)

**A2AExecutor (7 tests)**:
- Full RequestContext reference identity (not a copy)
- Hooks see `a2aRequestContext` with correct taskId/contextId
- Tools see it via BeforeToolCallEvent
- All lifecycle hooks share same reference through A2A path
- toJSON exclusion (no invocationState leakage)
- No bleed between sequential A2A requests
- Hook mutations don't corrupt RequestContext object

**Cancel paths (4 tests)**:
- invocationState survives `BeforeInvocationEvent.cancel`
- invocationState survives `BeforeModelCallEvent.cancel`
- Reference identity preserved through early cancel
- Cancelled invocation state doesn't bleed to next invocation

**Multiagent nodes (5 tests)**:
- AgentNode stream() preserves invocationState reference to inner agent
- Mutations from inner agent visible on outer invocationState
- Deeply nested objects propagate without corruption through Graph
- Empty invocationState defaults to `{}` when not provided
- Mutations from graph agent propagate back to caller

<details>
<summary>What Survived (all 16 tests passed)</summary>

- ✅ A2A executor correctly threads invocationState to all agent hooks
- ✅ Reference identity is preserved end-to-end (A2A → Agent → Hooks → Tools)
- ✅ toJSON correctly excludes invocationState from serialization
- ✅ Sequential A2A requests get isolated invocationState objects
- ✅ Cancel paths (BeforeInvocation, BeforeModelCall) preserve invocationState on AgentResult
- ✅ AgentNode properly forwards invocationState through stream() → handle() → inner agent
- ✅ Graph orchestration preserves reference identity across nodes
- ✅ Edge cases (null values, deeply nested objects, missing invocationState) all handled

</details>

</details>

**Recommendation**: @mkmeral — this is ready to approve. All prior feedback addressed, all adversarial tests pass, the implementation is clean and correct.

---
🤖 *AI agent response. [Strands Agents](https://github.com/strands-agents). Feedback welcome!*

**Review by mkmeral — APPROVED** [review]
